src/snowflake/connector/cache.py

```diff
@@ -0,0 +1,264 @@
+#
+# Copyright (c) 2012-2021 Snowflake Computing Inc. All rights reserved.
+#
+
+from __future__ import annotations
+
+import datetime
+from collections.abc import Iterator
+from threading import Lock
+from typing import Generic, TypeVar
+
+from typing_extensions import NamedTuple, Self
+
+now = datetime.datetime.now
+
+T = TypeVar("T")
+
+
+class CacheEntry(NamedTuple, Generic[T]):
+    expiry: datetime.datetime
+    entry: T
+
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+def is_expired(d: datetime.datetime) -> bool:
+    return now() >= d
+
+
+class SFDictCache(Generic[K, V]):
+    """A generic in-memory cache that acts somewhat like a dictionary.
+
+    Notes:
+    - Unlike normal dictionaries keys, values and items return list
```

Comment:
Maybe it's just me but I find this note not very fluent. How about:

`keys(), values() and items() return list materialized at call time, unlike normal dictionaries that return a view object`